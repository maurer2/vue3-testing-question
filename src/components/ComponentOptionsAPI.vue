<template>
  <section>
    <h1>Options API</h1>

    <ButtonComponent
      id="button-2a"
      @click="handleClick1"
    />
    <ButtonComponent
      id="button-2b"
      @click="handleClick2()"
    />
  </section>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import ButtonComponent from './ButtonComponent.vue';

export default defineComponent({
  name: 'ComponentOptionsAPI',
  components: {
    ButtonComponent,
  },
  data() {
    return {
      counter1: 0,
      counter2: 0,
    };
  },
  methods: {
    handleClick1() {
      this.counter1 += 1;
    },
    handleClick2() {
      this.counter2 += 1;
    },
  },
});

</script>

<style scoped>

</style>
