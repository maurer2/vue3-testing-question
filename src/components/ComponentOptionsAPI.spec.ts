// eslint-disable-next-line import/no-extraneous-dependencies
import { shallowMount } from '@vue/test-utils';
import ComponentOptionsAPI from './ComponentOptionsAPI.vue';

describe('Component Options Api', () => {
  let cmp: any;
  let spyHandleClick1: any;
  let spyRunOnMount: any;

  beforeEach(() => {
    spyHandleClick1 = jest.spyOn((ComponentOptionsAPI.methods as any), 'handleClick1');
    spyRunOnMount = jest.spyOn((ComponentOptionsAPI.methods as any), 'runOnMount');

    cmp = shallowMount(ComponentOptionsAPI, {});
  });

  afterEach(() => {
    cmp.unmount();

    jest.clearAllMocks();
    jest.resetModules();

    spyHandleClick1.mockClear();
  });

  it('renders', () => {
    expect(cmp.html()).toContain('Options API');
  });

  it('matches snapshot', () => {
    // https://github.com/eddyerburgh/jest-serializer-vue/issues/49
    // eslint-disable-next-line no-underscore-dangle
    expect(cmp.__app._container).toMatchSnapshot();
  });

  it('button1 triggers counter update function when clicked - spied before mount', async () => {
    await cmp.find('#button-2a').trigger('click');

    expect(spyHandleClick1).toHaveBeenCalled();
  });

  it('button1 triggers counter update function when clicked - spied on the fly', async () => {
    const spy = jest.spyOn(cmp.vm, 'handleClick1');

    await cmp.find('#button-2a').trigger('click');
    expect(spy).toHaveBeenCalled();
  });

  it('button2 triggers counter update function when clicked', async () => {
    const spy = jest.spyOn(cmp.vm, 'handleClick2');

    await cmp.find('#button-2b').trigger('click');
    expect(spy).toHaveBeenCalled();
  });

  it('button1 updates counter1', async () => {
    await cmp.find('#button-2a').trigger('click');
    expect(cmp.vm.counter1).toBe(1);
  });

  it('button2 updates counter2', async () => {
    await cmp.find('#button-2b').trigger('click');
    expect(cmp.vm.counter2).toBe(1);
  });

  it('runOnMount runs on mount', async () => {
    expect(spyRunOnMount).toHaveBeenCalled();
  });
});
