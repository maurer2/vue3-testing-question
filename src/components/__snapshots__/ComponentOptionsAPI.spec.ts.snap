// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Options Api matches snapshot 1`] = `
<div
  data-v-app=""
>
  <section>
    <h1>
      Options API
    </h1>
    <button-component-stub
      id="button-2a"
    />
    <button-component-stub
      id="button-2b"
    />
  </section>
</div>
`;
